freeipa/src/main/java/com/sequenceiq/freeipa/service/BootstrapService.java

```diff
@@ -129,4 +133,72 @@ private String getHostname(FreeIpa freeIpa, InstanceMetaData im) {
     private byte[] getStateConfigZip() throws IOException {
         return compressUtil.generateCompressedOutputFromFolders("salt-common", "freeipa-salt");
     }
+
+    private Set<Node> convertToNodes(FreeIpa freeIpa, Set<InstanceMetaData> instanceMetaDatas) {
+        return instanceMetaDatas.stream()
+                .map(im -> new Node(
+                        im.getPrivateIp(),
+                        im.getPublicIpWrapper(),
+                        im.getInstanceId(),
+                        im.getInstanceGroup().getTemplate().getInstanceType(),
+                        getHostname(freeIpa, im),
+                        freeIpa.getDomain(),
+                        im.getInstanceGroup().getGroupName()))
+                .collect(Collectors.toSet());
+    }
+
+    private BootstrapParams createBootstrapParams(Stack stack, Set<InstanceMetaData> instanceMetaDatas) {
+        BootstrapParams params = new BootstrapParams();
+        params.setCloud(stack.getCloudPlatform());
+        ImageEntity image = imageService.getByStack(stack);
+        params.setOs(image.getOs());
+        params.setSaltBootstrapFpSupported(true);
+        params.setRestartNeededFlagSupported(true);
+        instanceMetaDatas.stream()
+                .filter(instanceMetaData -> InstanceMetadataType.GATEWAY_PRIMARY.equals(instanceMetaData.getInstanceMetadataType()))
+                .findFirst()
+                .map(instanceMetaData -> instanceMetaData.getInstanceGroup().getTemplate().getInstanceType())
+                .ifPresent(s -> params.setMasterWorkerThreads(calcMasterWorkerThreads(stack, s)));
+        return params;
+    }
+
+    private int calcMasterWorkerThreads(Stack stack, String masterInstanceType) {
+        CloudPlatform cloudPlatform = CloudPlatform.valueOf(stack.getCloudPlatform());
+        Optional<PricingCache> pricingCache = getPricingCache(cloudPlatform);
+        int result = BootstrapParams.DEFAULT_WORKER_THREADS;
+        if (pricingCache.isPresent()) {
+            Credential credential = credentialService.getCredentialByEnvCrn(stack.getEnvironmentCrn());
+            ExtendedCloudCredential extendedCloudCredential = convertCredentialToExtendedCloudCredential(credential, cloudPlatform);
+            Optional<Integer> cpuCountForInstanceType =
+                    pricingCache.get().getCpuCountForInstanceType(stack.getRegion(), masterInstanceType, extendedCloudCredential);
+            if (cpuCountForInstanceType.isPresent()) {
+                result = cpuCountForInstanceType.get() - BootstrapParams.WORKER_THREAD_REDUCTION_COUNT;
+            }
+        }
+        return Math.max(BootstrapParams.DEFAULT_WORKER_THREADS, result);
+    }
```

Comment:
this logic is mostly duplicated, can't we have the common part somewhere where it's shared? also I'm not sure this is really required for freeipa at all

Reply:
Now it's only here.